plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/service/CodeWhispererService.kt

```diff
@@ -285,22 +300,42 @@ class CodeWhispererService(private val cs: CoroutineScope) : Disposable {
                                     }
                                     runInEdt {
                                         workerContexts.forEach {
-                                            states = processCodeWhispererUI(it, states)
+                                            processCodeWhispererUI(
+                                                sessionContext,
+                                                it,
+                                                ongoingRequests[currentJobId],
+                                                cs,
+                                                currentJobId
+                                            )
+                                            if (!ongoingRequests.contains(currentJobId)) {
+                                                cs.coroutineContext.cancel()
+                                            }
```

Comment:
andrew could you help me understand why isn't this in the format of
```
if (ongoingRequests.contains(currentJobId)) {
     processCodeWhispererUi()
}
```
but execute processCodeWhispererUi and cancel the coroutine later, what's the expected behavior here?

Reply:
so intially there's no InvocationContext in ongoingRequests, and then we contruct it in `initState` in `processCodeWhispererUi` and put it into the map. If after this function it's still null, then it means state creation failed due to e.g. all discarded suggestions or all empty suggestions.